lsp layer -- Configure for lsp-ui-peek OR xref/lsp navigation

cormacc opened the pull request:
Added a configuration option to specify a preference for lightweight xref/lsp navigation functions (for find definition/implementation etc.) or lsp-ui-peek. By default, lightweight functions bound under `m g` and lsp-ui-peek equivalents under `m p`. Setting a config var to either `simple` or `peek` eliminates the bindings under `m p prefix` and binds either the xref/lsp (simple) or lsp-ui-peek (peek) nav functions under `m g`.

Also added functions for derived layers to provide server-specific extensions in a manner consistent with the specified preference (i.e. using `lsp-find-custom`, `lsp-ui-peek-find-custom` or both as specified by the configuration variable).

Prompted by a number of comments on the original lsp-layer-improvements PR (#10486) and subsequently on the c-c++ lsp backend PR (#11242).

Opportunistically rebound the lsp-flycheck error jump list under `m g e` rather than `m g f` -- a better mnemonic -- e(rror) reflecting function, rather than f(lycheck) reflecting implementation. `m g f` better reserved for 'goto file at point` in derived language layers.

Renamed a layer configuration variable from `lsp-ui-remap-xref-keybindings` to `lsp-remap-keybindings` to clarify that it's a layer config variable rather than a config variable exposed by the lsp-ui package.

yyoncho commented on layers/+tools/lsp/funcs.el:
```diff
@@ -141,4 +159,56 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
+    (progn
+      (spacemacs//lsp-define-find-extension layer-name kind request extra)
+      (spacemacs//lsp-define-peek-extension layer-name kind request extra))
+    (progn
```
The comment is only for the second progn.

cormacc replied:
Thanks - corrected

cormacc commented on layers/+tools/lsp/funcs.el:
```diff
@@ -84,11 +81,32 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
+  (ecase lsp-navigation
+    ('simple (spacemacs//lsp-bind-simple-navigation-functions-for-mode mode "g"))
+    ('peek (spacemacs//lsp-bind-peek-navigation-functions-for-mode mode "g"))
+    ('both (progn
```
Thanks - corrected

cormacc commented:
Amended the commit to incorporate feedback @yyoncho and correct an error in the peek extension function definitions.

cormacc commented:
<<Amendment 1 21/09/18>>
    Added functions to facilitate keybindings in derived layers consistent
    with navigation style preference.
    
    <<Amendment 2 24/09/18>>
    Incorporated feedback from yyoncho
    Corrected error in peek binding extension definition
    
    <<Amendment 3 24/09/18>>
    Moved 'heirarchy' prefix/keybindings from 'SPC m h' to 'SPC m <g/p> h' for
    consistency with navigation style preference selections.
    Added 'members' prefix under 'SPC m <g/p> m'
    Moved 'lsp-ui-imenu' from 'SPC m g m' to 'SPC m g M'

cormacc commented:
@sdwolfz I've just amended this commit to revert an unintentional whitespace and bracing change in a function I hadn't intended to touch. Everything else here should be relatively risk free, in that all the changes relate to the default core keybinding functions and I'm pretty sure none of the other lsp-derived layers have started using this... yet (I'm an optimist :) ) -- i.e. the only client is my own PR against c-c++ layer

yyoncho commented:
@cormacc I read somewhere that @sdwolfz will be off for several weeks.

cormacc commented:
OK - thanks yyoncho

kvaneesh commented:
Any update on merging this?

cormacc commented:
I've rebased on develop tip and resolved the conflict (README formatting)

sdwolfz commented on layers/+tools/lsp/funcs.el:
```diff
@@ -87,11 +87,32 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
+    ('peek (spacemacs//lsp-bind-peek-navigation-functions-for-mode mode "g"))
+    ('both
+      (spacemacs//lsp-bind-simple-navigation-functions-for-mode mode "g")
+      (spacemacs//lsp-bind-peek-navigation-functions-for-mode mode "p"))))
```
Honestly I would prefer the `'both` option to not exist and only have the `'simple` or `'peek` options under `g`. This way `p` can be used by other functionality.

Another option would be lowercase for simple and uppercase for peek, but that might be confusing as well... What do you say?

cormacc replied:
I use both. `g` when I know where I'm going and `p` when I'm not sure -- the context given by the lsp-ui-peek overlay is very useful when using e.g. find-callers / references type lookup functions.
Though could certainly live with a different prefix (i.e. capital G as you suggest). So could go that route and either eliminate the preference config value altogether (i.e. always bind both sets of functions under g and G respectively) or rename the config var to `lsp-navigation-invert-bindings` or something -- when non-nil, bind the lsp-ui-peek variants under `g` and move the xref variants to `G`.

Any views from contributors to other lsp-derived layers?  @yyoncho , @jamartin9, @ljupchokotev, @ztlevi (based on a quick search for recent PRs mentioning lsp)

yyoncho replied:
@cormacc I use both as well, mainly xref ones so I prefer them as a default. In lsp java layer which is not synched with lsp base layer I am using `mgr` and `mgR` for UI peek but `mGr`  looks better.

cormacc replied:
OK - went with the simplest change -- just replaced 'p' with 'G'. Works better alright -- easy to remember. Rebased/amended/pushed.

cormacc replied:
@sdwolfz  Had done `git commit --amend` rather than `git commit -a --amend` <facepalm>, so change wasn't pushed earlier -- fixed now.

MaskRay commented on layers/+tools/lsp/funcs.el:
```diff
@@ -144,4 +165,55 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
+
+(defun spacemacs//lsp-define-peek-extension (layer-name kind request &optional extra)
+  (if extra
+    (defalias (spacemacs/lsp-get-peek-extension-name layer-name kind) `(lambda () ,(format "Peek %s" (symbol-name kind)) (interactive) (lsp-ui-peek-find-custom ',kind ,request ',extra)))
```
Heads-up. The extra `kind` parameter in `lsp-ui-peek-*` has been removed.

`lsp-find-custom` `lsp-ui-peek-find-custom` can be used the same way now.

cormacc replied:
Thanks for the heads-up Fangrui!
I've updated, amended, pushed. Nice that `lsp-find-custom` and `lsp-ui-peek-find-custom` have the same call syntax now. It's allowed me to simplify the chunk of elisp that defines the extensions considerably.

cormacc commented on layers/+tools/lsp/README.org:
```diff
@@ -42,74 +46,161 @@ If you add =lsp-*-enable= to major mode hooks for auto initialization of
+| prefix          | name             | functional area                                          |
+|-----------------+------------------+----------------------------------------------------------|
+| ~SPC m <g/p> h~ | hierarchy        | Heirarchy (i.e. call/inheritance hierarchy etc. )        |
+| ~SPC m <g/p> m~ | member hierarchy | Class/namespace members (functions, nested classes, vars |
```
Done

cormacc commented on layers/+tools/lsp/config.el:
```diff
@@ -9,8 +9,14 @@
+(defvar lsp-navigation 'both
+  "If `simple' binds lightweight navigation functions under `SPC m g'.
+If `peek' binds lsp-ui navigation functions under `SPC m g'.
+If `both', binds lightweight navigation functions under `SPC m g' and lsp-ui functions under `SPC m p'")
```
... and done

sdwolfz commented:
Thank you :heart:!
Cherry-picked into develop branch, you can safely delete your branch.

cormacc commented:
Haven't used that configuration option myself, but pretty sure it would only prevent the lsp server being loaded. To selectively load either rtags or lsp-cquery/ccls per project, you'd probably need to make quite a few changes to packages.el. Could maybe do something clever in the 'c-mode-common-hook, or with project / directory local variables? Not one of my own use cases, but feel to raise a PR

yyoncho commented:
@cormacc @sdwolfz I just realized that there is `spacemacs/set-leader-keys-for-minor-mode` thus we can configure all `lsp-mode` bindings in the base `LSP` layer. Like:

```
(spacemacs/set-leader-keys-for-minor-mode 'lsp-mode
    "ea"  'lsp-execute-code-action
    ...)
```

kvaneesh commented:
@cormacc also how is c-c++-lsp-project-whitelist suppose to interact with lsp-project-whitelist? Adding correct projects to whitelist and using @yyoncho suggestion to use keys-for-minor-mode should prevent us from enabling lsp specific key binding if the project is not in white list?

cormacc commented:
@kvaneesh c-c++-lsp-project-whitelist just gets mapped to ccls-project-whitelist or cquery-project-whitelist, depending on the selected backend, so I'd refer you to the upstream documentation. The spacemacs layer is just a bit of integration glue.

cormacc commented:
@yyoncho `set-leader-keys-for-minor-mode` sounds like a good option. Would this be enabled by default for all lsp layers? Would probably (almost definitely) result in some keybinding conflicts with other lsp-xxx layers initially, but maybe not a problem if the binding is 'weak' so to speak, i.e overridden by existing bindings in derived layers. Good in the long run for consistency of keybindings, encouraging buy-in etc., but might need a bit of managing in its introduction. @sdwolfz, what do you think?

yyoncho commented on layers/+tools/lsp/packages.el:
```diff
@@ -28,6 +28,7 @@
     :defer t
     :config
     (progn
+      (spacemacs|hide-lighter lsp-mode)
```
I guess this is bad merge, we need reenable the lighter one more time https://github.com/syl20bnr/spacemacs/pull/11283 :)

cormacc replied:
D'oh - sorry. I rebased this set of PRs so many times I obviously got sloppy :S

sdwolfz commented:
I think `spacemacs/set-leader-keys-for-minor-mode` gives warnings when you overwrite keybindings... Anyway I'm OK with this change, LSP support is still new and we have room for improvements and experimentation.
